### PR TITLE
xds: add CsdsService 

### PR DESCRIPTION
**Pull request by sergiitk**
Implements CsdsService, as defined in [gRFC A40: xDS Configuration Dump via Client Status Discovery Service in gRPC](https://github.com/grpc/proposal/blob/master/A40-csds-support.md).

cc @lidizheng

### CODE REVIEW

**Comment by lidizheng** [comment]
LGTM to https://github.com/grpc/grpc-java/pull/8008/commits/81b8a0a97efeae62f0bb52a817a24243054e6234, thanks for writing this test.

**Review by voidzcy — COMMENTED** [review]
In general it looks pretty good.

File: xds/src/test/java/io/grpc/xds/CsdsServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+
+  @RunWith(JUnit4.class)
+  public static class ServiceTests {
+    private static final CsdsService CSDS_SERVICE_MINIMAL =
+        new CsdsService(new FakeXdsClientPoolFactory(XDS_CLIENT_NO_RESOURCES));
+    private static final ClientStatusRequest REQUEST = ClientStatusRequest.getDefaultInstance();
+    private static final ClientStatusRequest REQUEST_INVALID =
+        ClientStatusRequest.newBuilder().addNodeMatchers(NodeMatcher.getDefaultInstance()).build();
+
+    @Rule public final GrpcServerRule grpcServerRule = new GrpcServerRule().directExecutor();
```
Comment (voidzcy) [inline xds/src/test/java/io/grpc/xds/CsdsServiceTest.java:106]: It's generally [favorable](https://github.com/grpc/grpc-java/blob/ad2b26436ccd64cb394127ebc0cd3323817dd8a3/testing/src/main/java/io/grpc/testing/GrpcServerRule.java#L38-L44) to use [`GrpcCleanupRule`](https://github.com/grpc/grpc-java/blob/master/testing/src/main/java/io/grpc/testing/GrpcCleanupRule.java). It's a newer thing that does everything GrpcServerRule can do plus a lot [extra](https://github.com/grpc/grpc-java/pull/4394#issue-184194598).

I am still fine with using GrpcServerRule.

**Review by voidzcy — APPROVED** [review]
Mostly LGTM.

File: xds/src/main/java/io/grpc/xds/CsdsService.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+// TODO(sergiitk): finish description, update since.
+/**
+ * The CSDS service provides information about the status of a running xDS client.
+ *
+ * @since 1.37.0
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/TODO")
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply xds/src/main/java/io/grpc/xds/CsdsService.java:58]: I just missed this todo. Tracking ticket created, will add this: https://github.com/grpc/grpc-java/issues/8016.
I think making it public still makes sense, so users may choose to use it independently. Also, won't `package-private` make it accessible to `AdminInterface`?

File: xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java [hunk 3]
```diff
@@ -79,6 +79,10 @@ protected int priority() {
   interface XdsClientPoolFactory {
     void setBootstrapOverride(Map<String, ?> bootstrap);
 
+    // TODO(sergiitk): docs, test?
+    @Nullable
+    ObjectPool<XdsClient> getXdsClientPoolOrNull();
+
     ObjectPool<XdsClient> getXdsClientPool() throws XdsInitializationException;
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java:86]: I agree, `getOrCreate()` or something else to indicate lazy-load is better. I'm OK with making this change in this PR, if that's ok with you.

[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsNameResolverProvider.java:86]: Sure, that would be great.

File: xds/src/test/java/io/grpc/xds/CsdsServiceTest.java (hunk 1 above)
Comment (sergiitk) [reply xds/src/test/java/io/grpc/xds/CsdsServiceTest.java:106]: I think, at least for as tests work now, `GrpcServerRule` is a really good fit. `CsdsServiceTest.ServiceTests` follow the same pattern - create server with channel to it, add service to the registry, make a request, test it. Changing `GrpcCleanupRule` would pretty much replicate this exact behavior provided by `GrpcServerRule` out of box.

File: xds/src/main/java/io/grpc/xds/CsdsService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/CsdsService.java:58]: Oh, you are right. I forgot that point.

**Pull request merged**